File: 20-valid-parentheses/20-valid-parentheses.cpp
class Solution {
public:
    bool isOpen(char& c){
        if(c=='(' || c=='{' || c=='[')
            return true;
        return false;
    }
    
    bool isValid(string s) {
        stack<char> bracket_stack;
        unordered_map<char,char> char_map{ {'(',')'}, {'{','}'}, {'[',']'} };
        if(s.size() == 0)
            return true;
        for(auto i: s){
            if(char_map.count(i)){
                bracket_stack.push(i);
            }
            else{
                if(bracket_stack.empty() || i != char_map[bracket_stack.top()]){
                    return false;
                }
                bracket_stack.pop();
            }
        }
        if(bracket_stack.empty())
            return true;
        return false;
    }
};

File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closers = {"]":"[","}":"{",")":"("}
        for i in s:
            if i in closers:
                if len(stack)==0 or stack[-1] != closers[i]:
                    return False
                else:
                    stack.pop()
            else:
                stack.append(i)
                
        return len(stack) == 0